### PR TITLE
Check duplicates in supply queue 109

### PR DESCRIPTION
**Pull request by MerlinEgalite**
Fixes https://github.com/cantinasec/review-morpho-blue-1/issues/109

### CODE REVIEW

File: src/MetaMorpho.sol [hunk 1]
```diff
@@ -289,6 +289,13 @@ contract MetaMorpho is ERC4626, ERC20Permit, Ownable2Step, Multicall, IMetaMorph
 
         for (uint256 i; i < length; ++i) {
             if (config[newSupplyQueue[i]].cap == 0) revert ErrorsLib.UnauthorizedMarket(newSupplyQueue[i]);
+
+            // Check duplicates.
+            for (uint256 j = i + 1; j < length; ++j) {
+                if (Id.unwrap(newSupplyQueue[i]) == Id.unwrap(newSupplyQueue[j])) {
```
Comment (Rubilmax) [inline src/MetaMorpho.sol:295]: `newSupplyQueue[i]` can be stored onto the stack

File: test/metamorpho_tests.tree [hunk 2]
```diff
@@ -221,7 +221,9 @@
         └── when newSupplyQueue.length <= MAX_QUEUE_SIZE
             ├── when some markets of newSupplyQueue have a zero cap
             │   └── revert with UnauthorizedMarket()
-            └── when all the markets of newSupplyQueue have non zero cap
+            ├── when some markets are duplicated
```
Comment (Rubilmax) [inline test/metamorpho_tests.tree:224]:
```suggestion
            ├── when there is at least a duplicate market
```

File: test/metamorpho_tests.tree [hunk 3]
```diff
@@ -221,7 +221,9 @@
...
+            │   └── revert with DuplicateMarket(id)
+            └── when all the markets of newSupplyQueue have non zero cap and are not duplicated
```
Comment (Rubilmax) [inline test/metamorpho_tests.tree:226]:
```suggestion
            └── when all the markets of newSupplyQueue have non zero cap and are unique
```

**Review by Rubilmax — CHANGES_REQUESTED** [review]

**Comment by Rubilmax** [comment]
For reference: the nested loop costs 3k in the case of a 5 markets supply queue

**Review by Jean-Grimal — APPROVED** [review]

File: src/MetaMorpho.sol [hunk 4]
```diff
@@ -288,7 +288,16 @@ contract MetaMorpho is ERC4626, ERC20Permit, Ownable2Step, Multicall, IMetaMorph
...
-            if (config[newSupplyQueue[i]].cap == 0) revert ErrorsLib.UnauthorizedMarket(newSupplyQueue[i]);
+            Id id = newSupplyQueue[i];
+
+            if (config[id].cap == 0) revert ErrorsLib.UnauthorizedMarket(id);
+
+            // Check duplicates.
+            for (uint256 j = i + 1; j < length; ++j) {
+                if (Id.unwrap(id) == Id.unwrap(newSupplyQueue[j])) {
+                    revert ErrorsLib.DuplicateMarket(id);
+                }
```
Comment (Rubilmax) [inline src/MetaMorpho.sol:299]:
```suggestion
                if (Id.unwrap(id) == Id.unwrap(newSupplyQueue[j])) revert ErrorsLib.DuplicateMarket(id);
```

**Review by Rubilmax — APPROVED** [review]
The implementation works, but it costs a lot

**Review by Rubilmax — CHANGES_REQUESTED** [review]
Here's my reasoning: https://github.com/cantinasec/review-morpho-blue-1/issues/109#issuecomment-1790698623

File: src/MetaMorpho.sol (hunk 4 above)
Comment (MerlinEgalite) [reply src/MetaMorpho.sol:299]: Indeed the compiled does not complain

**Comment by MerlinEgalite** [comment]
> For reference: the nested loop costs 3k (5%) in the case of a 5 markets supply queue and 116k (60%) in the case of a 30 markets supply queue

3k is ok.

I agree that 116k is a lot but I'm not sure we this function we'll be trigger several times a day tbh

**Comment by MerlinEgalite** [comment]
We acknowledge this issue because:
- The gas cost is important
- The the curator can fix a rogue allocator by disabling the allocator and setting a new supply queue

**Pull request closed without merging**